hashmap/hashmap_test.mbt

```diff
@@ -0,0 +1,217 @@
+  m.set("a", 1)
+  m.set("ab", 2)
+  m.set("bc", 2)
+  m.set("cd", 2)
+  m.set("abc", 3)
+  m.set("abcdef", 6)
+  m.remove("ab")
+  @assertion.assert_eq(m.size(), 5)?
+  @assertion.assert_eq(
+    m.debug_entries(),
+    "_,(0,a,1),(0,bc,2),(1,cd,2),(1,abc,3),_,(0,abcdef,6),_",
+  )?
+}
+
+test "remove_unexist_key" {
+  let m : HashMap[String, Int] = HashMap::new(hasher=Some(fn(k) { k.length() }))
+  m.set("a", 1)
+  m.set("ab", 2)
+  m.set("abc", 3)
+  m.remove("d")
+  @assertion.assert_eq(m.size(), 3)?
+  @assertion.assert_eq(
+    m.debug_entries(),
+    "_,(0,a,1),(0,ab,2),(0,abc,3),_,_,_,_",
+  )?
+}
+
+test "size" {
+  let m : HashMap[String, Int] = HashMap::new()
+  @assertion.assert_eq(m.size(), 0)?
+  m.set("a", 1)
+  @assertion.assert_eq(m.size(), 1)?
+}
+
+test "is_empty" {
+  let m : HashMap[String, Int] = HashMap::new()
+  @assertion.assert_eq(m.is_empty(), true)?
+  m.set("a", 1)
+  @assertion.assert_eq(m.is_empty(), false)?
+  m.remove("a")
+  @assertion.assert_eq(m.is_empty(), true)?
+}
+
+test "iter" {
+  let m : HashMap[String, Int] = HashMap::[("a", 1), ("b", 2), ("c", 3)]
+  let mut sum = 0
+  m.iter(fn(_k, v) { sum += v })
+  @assertion.assert_eq(sum, 6)?
+}
+
+test "iteri" {
+  let m : HashMap[String, Int] = HashMap::[("a", 1), ("b", 2), ("c", 3)]
+  let mut sum = 0
+  let mut s = ""
+  m.iteri(
+    fn(i, _k, v) {
+      s += i.to_string()
+      sum += v
+    },
+  )
+  @assertion.assert_eq(s, "012")?
+  @assertion.assert_eq(sum, 6)?
+}
+
+test "clear" {
+  let m : HashMap[String, Int] = HashMap::[("a", 1), ("b", 2), ("c", 3)]
+  m.clear()
+  @assertion.assert_eq(m.size, 0)?
+  @assertion.assert_eq(m.capacity, 8)?
+  for i = 0; i < m.capacity; i = i + 1 {
+    @assertion.assert_is(m.entries[i], Empty)?
+  }
+}
+
+test "grow" {
+  let m : HashMap[String, Int] = HashMap::new(hasher=Some(fn(k) { k.length() }))
+  m.set("C", 1)
+  m.set("Go", 2)
+  m.set("C++", 3)
+  m.set("Java", 4)
+  m.set("Scala", 5)
+  m.set("Julia", 5)
+  @assertion.assert_eq(m.size, 6)?
+  @assertion.assert_eq(m.capacity, 8)?
+  m.set("Cobol", 5)
+  @assertion.assert_eq(m.size, 7)?
+  @assertion.assert_eq(m.capacity, 16)?
+  m.set("Python", 6)
+  m.set("Haskell", 7)
+  m.set("Rescript", 8)
+  @assertion.assert_eq(m.size, 10)?
+  @assertion.assert_eq(m.capacity, 16)?
+  @assertion.assert_eq(
+    m.debug_entries(),
+    "_,(0,C,1),(0,Go,2),(0,C++,3),(0,Java,4),(0,Scala,5),(1,Julia,5),(2,Cobol,5),(2,Python,6),(2,Haskell,7),(2,Rescript,8),_,_,_,_,_",
+  )?
+}
+
+test "as_iter" {
+  let buf = Buffer::make(20)
+  let map = HashMap::[(1, "one"), (2, "two"), (3, "three")]
+  map.as_iter().iter(
+    fn(e) {
+      let (k, v) = e
+      buf.write_string("[\(k)-\(v)]")
+    },
+  )
+  inspect(buf, content="[1-one][2-two][3-three]")?
+  buf.reset()
+  map.as_iter().take(2).iter(
+    fn(e) {
+      let (k, v) = e
+      buf.write_string("[\(k)-\(v)]")
+    },
+  )
+  inspect(buf, content="[1-one][2-two]")?
+}
```

Comment:
Note that there is a missing blank line at the end here. Did you format it?

Reply:
Done.